mmgen/models/architectures/biggan/generator_discriminator_deep.py

```diff
@@ -0,0 +1,635 @@
+                Defaults to False.
+
+        Returns:
+            torch.Tensor | dict: If not ``return_noise``, only the output image
+                will be returned. Otherwise, a dict contains ``fake_img``,
+                ``noise_batch`` and ``label`` will be returned.
+        """
+        if isinstance(noise, torch.Tensor):
+            assert noise.shape[1] == self.noise_size
+            assert noise.ndim == 2, ('The noise should be in shape of (n, c), '
+                                     f'but got {noise.shape}')
+            noise_batch = noise
+        # receive a noise generator and sample noise.
+        elif callable(noise):
+            noise_generator = noise
+            assert num_batches > 0
+            noise_batch = noise_generator((num_batches, self.noise_size))
+        # otherwise, we will adopt default noise sampler.
+        else:
+            assert num_batches > 0
+            noise_batch = torch.randn((num_batches, self.noise_size))
+
+        if self.num_classes == 0:
+            label_batch = None
+
+        elif isinstance(label, torch.Tensor):
+            assert label.ndim == 1, ('The label shoube be in shape of (n, )'
+                                     f'but got {label.shape}.')
+            label_batch = label
+        elif callable(label):
+            label_generator = label
+            assert num_batches > 0
+            label_batch = label_generator((num_batches, ))
+        else:
+            assert num_batches > 0
+            label_batch = torch.randint(0, self.num_classes, (num_batches, ))
+
+        # dirty code for putting data on the right device
+        noise_batch = noise_batch.to(get_module_device(self))
+        if label_batch is not None:
+            label_batch = label_batch.to(get_module_device(self))
+            class_vector = self.shared_embedding(label_batch)
+        else:
+            class_vector = None
+
+        # If 'concat noise', concat class vector and noise chunk
+        if self.concat_noise:
+            if class_vector is not None:
+                z = torch.cat([noise_batch, class_vector], dim=1)
+                y = z
+        elif self.num_classes > 0:
+            z = noise_batch
+            y = class_vector
+        else:
+            z = noise_batch
+            y = noise_batch
+
+        # First linear layer
+        x = self.noise2feat(z)
+        # Reshape
+        if self.reshape_style == 'tf':
+            x = x.view(x.size(0), self.input_scale, self.input_scale, -1)
+            x = x.permute(0, 3, 1, 2).contiguous()
+        else:
+            x = x.view(x.size(0), -1, self.input_scale, self.input_scale)
+        # Loop over blocks
+        for idx, conv_block in enumerate(self.conv_blocks):
+            # Second inner loop in case block has multiple layers
+            if isinstance(conv_block, SelfAttentionBlock):
+                x = conv_block(x)
+            else:
+                x = conv_block(x, y)
```

Comment:
In some cases, y will be `noise_batch`. Will it work correctly in these cases?

Reply:
In fact, my unit tests cover this case. However, I will change the logic when `num_classes==0`. In other words, `y` would be `None` when `num_classes` is set to 0.

Reply:
Ok, thanks.